### PR TITLE
Add Hyrax multilinear PCS

### PR DESCRIPTION
**Pull request by Antonio95**
## Description

This PR implements the Hyrax polynomial commitment scheme: a multilinear PCS based on the hardness of the discrete logarithm problem introduced as part of the Hyrax zkSNARK in [this 2017 article](https://eprint.iacr.org/2017/1132.pdf).

The PCS described therein is interactive. When implementing the Fiat-Shamir transform, [this paper](https://eprint.iacr.org/2023/691) was consulted.

### Modification note

In the PCS contained in the cited article, the verifier never learns the actual evaluation of the polynomial at the requested point, but is instead convinced that a previously received Pedersen commitment is indeed a commitment to said evaluation - this is what the SNARK proposed therein necessitates. However, the Arkworks framework requies the verifier to actually learn that value, which is why we have added the opening of the commitment at the end of the protocol. This likely does not result in an optimal non-hiding PCS, but we feel it is the most faithful adaptation of the original PCS that can be implemented with the current restrictions.

### Future optimisations

Some natural optimisations to the scheme which are not part of the current PR, but would make sensible follow-up work, are the following:
- Deal with the modification described above: either modify the PCS trait to encompass hiding PCSs (in terms of the actual   evaluation, not only the polynomial), or turn this scheme into a non-hiding one by removing unnecessary work (which would   probably involve non-trivial theoretical work).
- Add parallelisation. There is at least one natural place where parallelisation could bring performance gains: in essence, the prover commits to the polynomial by expressing it as an evaluation matrix and Pederson-multi-committing to each row. Each of this commitments can be computed independently from the rest, and therefore, in parallel. It is still to be seen how much of an improvement this would entail, since each Pederson multi-commitment boils down to a multi-exponentiation and this operation is itself parallelised.
- Due to the homomorphic nature of Pedersen commitments, it is likely some of the following methods can be designed more efficiently than their default implementations: `batch_open`, `batch_check`, `open_combinations`, `check_combinations`. This is not discussed in the reference article, but the IPA and KZG modules might be a good starting point.
- On a related note to the previous point, there might be a more efficient way to open several polynomials at a single point (this is the functionality of the `open` method) than the currently implemented technique, where only the computation of the vectors `L` and `R` is shared across polynomials.
- The cited article proposes an optimisation in the section _Reducing the cost of proof-of-dot-prod_. It allows for non-square matrices (and hence removes the requirement for the number of variables to be even) and introduces a tradeoff between proof size and verifier time. It is probably worth pursuing.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by mmagician** [comment]
This PR relies on https://github.com/arkworks-rs/algebra/pull/691, so we temporarily expect CI to fail until that's merged.

**Comment by Pratyush** [comment]
Sorry for the late update on this, but happy to merge this as-is, once it's updated wrt master.

**Comment by autquis** [comment]
It is ready for review and to be merged.
@Pratyush If you are merging these three PRs, please do this PR first. Then, I will resolve the conflicts of the other two and let you know when they are ready.

File: bench-templates/src/lib.rs
Comment (Pratyush) [inline bench-templates/src/lib.rs:33]: Since we're adding quite a few arguments here, some comments would be helpful =)

Comment (autquis) [reply bench-templates/src/lib.rs:36]: Done.
However, I feel like this `bench-templates` can be improved. It is exclusive to multilinear polynomials at the moment. (Since `max_degree` is set to `num_vars`)

Comment (autquis) [reply bench-templates/src/lib.rs:36]: A gentle reminder on this @Pratyush

File: poly-commit/src/hyrax/mod.rs
Comment (Pratyush) [inline poly-commit/src/hyrax/mod.rs:96]: Is this always a hiding commitment?

File: poly-commit/src/hyrax/mod.rs
Comment (Pratyush) [inline poly-commit/src/hyrax/mod.rs:116]: This requires an extra allocation. A better way to compute this would be just do an MSM with `com_key`, and then add in the product `h * r` separately.

This latter addition can also be pulled out of the `pedersen_commit` (which means that you don't need to pass in `rng` or `r`)

File: poly-commit/src/hyrax/mod.rs
Comment (Pratyush) [inline poly-commit/src/hyrax/mod.rs:379]:
```suggestion
                .zip(&state.randomness)
```

File: poly-commit/src/hyrax/mod.rs
Comment (Pratyush) [inline poly-commit/src/hyrax/mod.rs:363]: This will be quite slow in practice. A faster way would be compute  the hash of the generators, store this in the commitment key and absorb that hash here.

File: poly-commit/src/hyrax/mod.rs
Comment (Pratyush) [inline poly-commit/src/hyrax/mod.rs:499]: Ditto

File: poly-commit/src/utils.rs
Comment (Pratyush) [inline poly-commit/src/utils.rs:21]: `ark-std` now contains a `serialize_to_vec` macro that you might want to use instead of this.

File: poly-commit/src/hyrax/mod.rs
Comment (autquis) [reply poly-commit/src/hyrax/mod.rs]: Done!

File: poly-commit/src/hyrax/mod.rs
Comment (autquis) [reply poly-commit/src/hyrax/mod.rs]: Got removed!

File: poly-commit/src/utils.rs
Comment (autquis) [reply poly-commit/src/utils.rs]: Done!

File: poly-commit/src/utils.rs
Comment (autquis) [reply poly-commit/src/utils.rs]: Also, done! :) I used `cfg_into_iter` for the Range, though.

File: poly-commit/src/hyrax/mod.rs
Comment (autquis) [reply poly-commit/src/hyrax/mod.rs]: I think there was a bug too. `r` was pushed into `scalars_ext` but `scalars_ext` was never used.

File: poly-commit/src/hyrax/mod.rs
Comment (autquis) [inline poly-commit/src/hyrax/mod.rs:254]: I am confused about this @Antonio95 . Here, we build a row-major matrix from `m`. But `m` is a column-major representation of evaluations. So which is what? Moreover, it is unclear to me why we need `m` to be column-major in the first place.

File: poly-commit/src/hyrax/mod.rs
Comment (autquis) [reply poly-commit/src/hyrax/mod.rs]: I will let the expert talk! :D @Antonio95

File: poly-commit/src/hyrax/mod.rs
Comment (autquis) [reply poly-commit/src/hyrax/mod.rs]: I was not sure which part was the slow part, serializing or absorbing. I assumed the latter. Please check https://github.com/arkworks-rs/poly-commit/pull/130/commits/ac4a14c5e7a3486d845229e0ef529ed64457648e and let me know if it was what you meant

File: poly-commit/src/hyrax/mod.rs
Comment (Antonio95) [reply poly-commit/src/hyrax/mod.rs]: Hey there! So, in [its published form](https://eprint.iacr.org/2017/1132) (subsection 6.1), this PCS is always hiding indeed: it hides both the polynomial coefficients (by using an extra point in the Pedersen commitment key, as well as an additional random coefficient for each commitment) and the evaluation of the polynomial at the requested point, which is never learnt by the verifier.

One of the authors (J. Thaler) has mentioned a non-hiding version in private discussions (not involving me). I imagine this entails using non-hiding Pedersen commitments, but some more substantial changes would need to be made to the dot product argument (part of the opening and verification methods) as well. This has, as far as I know, never appeared in print.

File: poly-commit/src/hyrax/mod.rs
Comment (Antonio95) [reply poly-commit/src/hyrax/mod.rs:254]: Hi Hossein! The method `flat_to_matrix_column_major` converts a flat list L into a matrix M interpreting the former as a column-first flattening of the latter. But, independently from that, there is the question of how that M is returned, i.e. how it is represented. As you can see in the [function documentation](https://github.com/HungryCatsStudio/poly-commit/blob/main/src/hyrax/utils.rs#L7), M is returned as a list of rows - and that is precisely the representation that the constructor `new_from_rows` expects.

I feel like the example in the documentation is clearer than the explanation therein - if you think it is necessary, we can fix the latter.

Let me know if that makes sense.

Comment (autquis) [reply poly-commit/src/hyrax/mod.rs:254]: Hi! Thanks for the explanation!
I think the issue is that `flat_to_matrix_column_major` returns vectors that are the columns of the matrix, but `new_from_rows` crafts a new matrix from the vectors of the rows. This could be a problem since we have matrix multiplication in
https://github.com/HungryCatsStudio/poly-commit/blob/ac4a14c5e7a3486d845229e0ef529ed64457648e/poly-commit/src/hyrax/mod.rs#L354
It seems to me that by this process, `t` in the line above is the transpose of `m`.


Also, I think it would be better if we do not use columns major for performance reasons. `flat_to_matrix_column_major` is a lot of work just for representing the matrix from row-major to column-major.

Comment (Antonio95) [reply poly-commit/src/hyrax/mod.rs:254]: Hey Hossein! As I mentioned in the first paragraph of my previous message, `flat_to_matrix_column_major` returns a list whose sublists are the *rows* of the matrix. Consider the example from the documentation:

Screenshot 2024-01-24 at 08 52 25

The "column-major" term in the function name only refers to how the flat input list is interpreted, not to the representation in which the matrix is returned. In particular,
- At no in the execution of `commit`/`open` do these functions construct a list whose sublists are columns of a matrix.
- At no point does the transpose of the polynomial coefficient matrix occur.

As an additional hint that things are done correctly (of course, insufficient on its own), note the commitment, opening and verification methods all work correctly with each other. If you transpose the matrix in either `commit` or `open`, the process will break (as expected).

Comment (autquis) [reply poly-commit/src/hyrax/mod.rs:254]: Hi Antonio!
Thanks for the explanation! The source of confusion was that I thought we should interpret the evaluation table as a row-major matrix rather than a column-major (as in Brakedown).

Comment (Antonio95) [reply poly-commit/src/hyrax/mod.rs:254]: I agree it can be a bit confusing! Is it clear now, then? Are you happy to leave that part of the code as it is?

**Comment by Antonio95** [comment]
I am happy with the current state of things! Let us see if @Pratyush would like any further tweaks regarding the open threads  of this discussion.

**Comment by autquis** [comment]
A gentle reminder @Pratyush :)

**Comment by autquis** [comment]
Gentle reminder @Pratyush. (It would be much simpler if we go ahead and merge, and then, we can always improve the code.)

**Review by Pratyush — APPROVED** [review]

File: poly-commit/src/hyrax/mod.rs
Comment (autquis) [inline poly-commit/src/hyrax/mod.rs:501]: Why is this check called "second" if it is not the second check? Maybe remove "first" and "second" and only refer to the equation number?

Comment (Antonio95) [reply poly-commit/src/hyrax/mod.rs:501]: They are commented as "first check from the paper" and "second check from the paper" (not just as "first" and "second") because that's the order they appear in the paper - which is true regardless of our implementation. It just so happens that it's (potentially) more efficient to make them in the opposite order for early returns. I don't think "first check from the paper" and "second check from the paper" can lead to any confusion, right?

Comment (autquis) [reply poly-commit/src/hyrax/mod.rs:501]: Hmm, I see. I thought the equation number would be enough. But true, it is not really confusing.

Comment (autquis) [reply poly-commit/src/hyrax/mod.rs:501]: My commit was unsigned, I had to rebase and repush your commit.

Comment (Antonio95) [reply poly-commit/src/hyrax/mod.rs:501]: No problem about the rebasing :) thanks for the fix

**Review by Pratyush — APPROVED** [review]

**Comment by autquis** [comment]
Do you know what blocks this PR from getting merged? @mmagician

**Comment by Pratyush** [comment]
Seems like the same build failure here

**Comment by autquis** [comment]
This PR is ready now. After this getting merged, I will update #131 and #132.

**Pull request merged**
